audit-checklists/PolygonPoS.md

```diff
@@ -0,0 +1,226 @@
+
+### `onStateReceive` Permissions
+
+```
+function onStateReceive(uint256 id, bytes calldata message) external only(STATE_SYNCER_ROLE) {
+    _processMessageFromRoot(message);
+}
+```
+Only the address `0x000...1001` (Polygon system address) can call `onStateReceive`. This is enforced via `only(STATE_SYNCER_ROLE)`:
+
+```
+## BaseChildTunnel.sol
+
+function onStateReceive(uint256, bytes calldata message) external only(STATE_SYNCER_ROLE) {
+        _processMessageFromRoot(message);
+    }
+
+    /**
+     * @notice Emit message that can be received on Root Tunnel
+     * @dev Call the internal function when need to emit message
+     * @param message bytes message that will be sent to Root Tunnel
+     * some message examples -
+     *   abi.encode(tokenId);
+     *   abi.encode(tokenId, tokenMetadata);
+     *   abi.encode(messageType, messageData);
+     */
+    function _sendMessageToRoot(bytes memory message) internal {
+        emit MessageSent(message);
+    }
+
+    /**
+     * @notice Process message received from Root Tunnel
+     * @dev function needs to be implemented to handle message as per requirement
+     * This is called by onStateReceive function.
+     * Since it is called via a system call, any event will not be emitted during its execution.
+     * @param message bytes message that was sent from Root Tunnel
+     */
+    function _processMessageFromRoot(bytes memory message) virtual internal;
+```
+- The `_processMessageFromRoot` function is virtual and must be implemented in inheriting contracts. Avoid unbounded loops or expensive external calls inside _processMessageFromRoot.
+
+Example bugs:
+
+1.  **Spoofed Data**: If `_processMessageFromRoot` doesn’t validate message contents, malicious inputs could be passed my attackers.  
+    Example: A protocol decoded `message` as `(address user, uint amount)` but forgot to check `amount > 0`, allowing free mints.
+    
+2.  **Replay Attacks**: Without state IDs/nonces, the same message could be processed twice.  
+
+3.  **Gas Limits**: Complex logic in `_processMessageFromRoot` might fail during gas spikes.
+
+
+## L2 → L1 Withdrawal Risks on Polygon PoS
+
+Polygons withdrawal system relies on checkpoints submitted to Ethereum. When users withdraw funds, validators batch the transactions into Merkle trees and submit the roots to Ethereum’s `RootChainManager`.
+
+The withdrawal path includes:
+
+-   Event emission on Polygon (for example, `Withdraw()`)
+    
+-   Inclusion of that block in a checkpoint
+    
+-   Finalization on Ethereum
+    
+-   Merkle proof verification on `RootChainManager`
+
+The key part of the withdrawal flow is validating Merkle proofs of inclusion:
+
+```
+## RootChainManager.sol
+_checkBlockMembershipInCheckpoint(...);
+```
+> Always use Polygon’s audited [MerklePatriciaProof](https://github.com/0xPolygon/pos-contracts/tree/main/contracts/common/lib) library and validate all proof components.
```

Comment:
Always point to a specific commit, i.e. https://github.com/0xPolygon/pos-contracts/tree/ddad751/contracts/common/lib